petastorm/pytorch.py

```diff
@@ -423,3 +425,160 @@ def __enter__(self):
+
+        :param reader: petastorm Reader instance
+        :param batch_size: the number of items to return per batch; factored into the len() of this reader
+        :param transform_fn: an optional callable to convert batches from the reader to PyTorch tensors
+        :param shuffling_queue_capacity: Queue capacity is passed to the underlying :class:`tf.RandomShuffleQueue`
+          instance. If set to 0, no shuffling will be done.
+        :param async_batch_queue_capacity: Number of batches to be produced in advance. If async_batch_queue_capacity
+            == 0, shuffle buffer will generate batches based on request. If async_batch_queue_capacity > 0, shuffle
+            work is done by an asynchronous thread to generate batches in advance. It is suggested to set this value
+            to be non-zero if shuffling is bottleneck for dataloader.
+        :param inmemory_cache_all: This is an boolean that indicates whether the data should be
+            cached in memory or not. When this cache is enabled, expect smaller than requested
+            batch size on the epoch boundaries. Enabling this cache also requres that shuffling_queue_capacity to be
+            large enough to hold all data.
+        :param num_epochs: Number of epochs to load when in memory cache is enabled. If
+            set to None, loader will loads batches indefinitely.
+        :param async_batch_queue_capacity: Capacity of queue to store shuffled batches. Shuffle work is done by an
+            asynchronous thread to produce batches into a queue.
+        """
+        super(AsyncBatchedDataLoader, self).__init__(reader, batch_size=batch_size, transform_fn=transform_fn,
+                                                     shuffling_queue_capacity=shuffling_queue_capacity,
+                                                     inmemory_cache_all=inmemory_cache_all,
+                                                     num_epochs=num_epochs)
+        self._async_batch_queue = Queue(async_batch_queue_capacity)
+        self._finished_event = Event()
+        self._shuffle_thread = None
+        self._started = False
+
+    def _worker(self):
+        for batch in self._iter_impl_worker():
+            if self._finished_event.is_set():
+                break
+            self._async_batch_queue.put(batch)
+        else:
+            self._async_batch_queue.put(None)
+
+    def _iter_impl(self):
+        # Start async shuffle thread
+        self._shuffle_thread = Thread(target=self._worker)
+        self._started = True
+        self._shuffle_thread.start()
+        while True:
+            batch = self._async_batch_queue.get()
+            if batch is None:
+                # None is last data in async batch queue, which means that async shuffle thread
+                # has reached the end of data.
+                self._started = False
+                break
+            yield batch
+
+        self._shuffle_thread.join()
+
+    def _iter_impl_worker(self):
```

Comment:
Appears that _iter_impl_worker has exactly the same implementation as its baseclass's: `BatchedDataLoader` `_iter_impl` implementation. Can we avoid this copy/paste somehow?

Reply:
`_iter_impl()` is from parent class that base dataloader uses to iterate, maybe I can do super. _iter_impl() for `thread` function, I will try.